File: 1000-minimum-cost-to-merge-stones/1000-minimum-cost-to-merge-stones.js
/**
 * @param {number[]} stones
 * @param {number} k
 * @return {number}
 */
var mergeStones = function(stones, k) {
  const n = stones.length;
  if ((n - 1) % (k - 1)) return -1;
  
  const dp = Array.from({ length: n }, () => new Array(n).fill(0));
  const preSum = stones.reduce((acc, stone) => (acc.push(acc.at(-1) + stone), acc), [0]);
  
  for (let i = n - 1; i >= 0; i--) {
    for (let j = i + 1; j < n; j++) {
      dp[i][j] = Infinity;
      for (let mid = i; mid < j; mid += k - 1) 
        dp[i][j] = Math.min(dp[i][j], dp[i][mid] + dp[mid + 1][j]);
      
      if ((j - i) % (k - 1) === 0)
        dp[i][j] += preSum[j + 1] - preSum[i];
    }
  }
  
  return dp[0][n - 1];
};

File: 1000-minimum-cost-to-merge-stones/1000-minimum-cost-to-merge-stones.py
class Solution:
    def mergeStones(self, stones: List[int], k: int) -> int:
        n = len(stones)
        if (n - 1) % (k - 1):
            return -1
        prefix = list(accumulate(stones, initial=0))
        @cache
        def dp(i:int, j:int):
            if i == j:
                return 0
            res = min(dp(i, mid) + dp(mid + 1, j) for mid in range(i, j, k - 1))
            if (j - i) % (k - 1) == 0:
                res += prefix[j + 1] - prefix[i]
            return res
        
        return dp(0, n - 1)